Add lookups to Marlin

jules opened the pull request.

jules committed:
Add lookup table base

jules committed:
Add lookup functionality to R1CS abstraction

jules committed:
Incorporate rudimentary lookup identity test in sumcheck

jules committed:
Clean up leftovers in z_m sum

jules committed:
Add rudimentary version of revised sumcheck

jules committed:
Compute and interpolate z_c, simplify lincheck sumcheck to extract rowcheck

ljedrz commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -98,7 +98,7 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         let sumcheck_lhs = Self::calculate_lhs(&state, t, summed_z_m, z, *alpha);
 
-        debug_assert!(
+        assert!(
```
note: this was a `debug_assert` presumably for performance reasons; might make sense to revert back to it, unless `evaluate_over_domain_by_ref` became fast enough in the meantime

ljedrz commented on algorithms/src/snark/plookup/table.rs:
```diff
@@ -0,0 +1,40 @@
+        Self(vec![vec![F::zero(); cols]; rows])
+    }
+
+    pub fn push(&mut self, row: usize, col: usize, element: F) {
```
this name might suggest that the table grows in the process; I'd suggest something like `fill` instead

jules committed:
Add extra PIOP round and insert updated rowcheck

jules committed:
Minor fixes

jules committed:
Add evaluations and witness commitments for z_c, selectors and query poly

jules committed:
Combined rowcheck and lincheck LC now sums to zero

ljedrz commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -189,26 +189,73 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+                let z_a_i = witness_label("z_a", i);
+                LinearCombination::new(z_a_i.clone(), [(F::one(), z_a_i)])
+            })
+            .collect::<Vec<_>>();
```
do these need to be collected? seems like they get iterated over immediately afterwards

ljedrz replied:
ah, I see that they are later reused too; there's still a possibility the allocations can be avoided, but it's not as trivial, so might be best to leave it for later :+1:

jules committed:
All checks hold now in proving and verification

This commit also fixes missing fields in serialization

jules committed:
Remove unused batch computations

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs:
```diff
@@ -74,11 +87,24 @@ impl<F: Field> ConstraintSystem<F> {
 impl<F: Field> CS<F> for ConstraintSystem<F> {
     type Root = Self;
 
+    fn add_lookup_table(&mut self, lookup_table: LookupTable<F>) -> Result<(), SynthesisError> {
```
Right now this just replaces the existing lookup table; I presume the reason it's named this way is because in the future we want to support multiple lookup tables? If so we should add a TODO comment or something.

jules replied:
yeah thats right! just left it like this for testing mostly - will annotate it

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs:
```diff
@@ -27,6 +34,9 @@ pub(crate) struct ConstraintSystem<F: Field> {
     pub(crate) num_public_variables: usize,
     pub(crate) num_private_variables: usize,
     pub(crate) num_constraints: usize,
+    pub(crate) lookup_table: Option<LookupTable<F>>,
+    pub(crate) s_mul: Vec<F>,
+    pub(crate) s_lookup: Vec<F>,
```
Instead of `s_mul` and `s_lookup` which are `num_constraints`-sized vectors containing `F`-sized elements, we can save memory as well as generalize to a future with multiple lookup tables as follows:
```rust
...
    pub(crate) mul_constraints: HashSet<ConstraintIndex>,
    pub(crate) lookup_constraints: Vec<LookupConstraints<F>>,
}

pub type ConstraintIndex = usize;
pub struct LookupConstraints<F> {
    table: LookupTable<F>,
    indices: HashSet<ConstraintIndex>
}
```

If we know we're going to have at most `u32::MAX` constraints, we can even set `ConstraintIndex = u32` to save space further.

Pratyush replied:
We can construct the selector polynomials using this compactified information later =)

Pratyush replied:
Edited the original comment to change the `Vec<ConstraintIndex>` to a `HashSet<ConstraintIndex>`.

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs:
```diff
@@ -124,6 +150,21 @@ impl<F: Field> CS<F> for ConstraintSystem<F> {
+        self.s_lookup.push(F::zero());
+    }
+
+    fn lookup(&mut self, val: LinearCombination<F>) -> Result<Variable, SynthesisError> {
```
This is a unary lookup, right? I think this suffices for Sinsemillia, but we should also support binary lookups `lookup(a, b) = c`, since those are useful for things like XOR, ANDs, etc.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/constraint_system.rs:
```diff
@@ -16,14 +16,24 @@
     pub(crate) num_public_variables: usize,
     pub(crate) num_private_variables: usize,
     pub(crate) num_constraints: usize,
+    pub(crate) lookup_table: Option<LookupTable<F>>,
+    pub(crate) s_mul: Vec<F>,
+    pub(crate) s_lookup: Vec<F>,
```
Same suggestion from the `indexer`

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -74,36 +74,53 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         assert_eq!(private_variables.len(), batch_size);
         let mut r_b_s = Vec::with_capacity(batch_size);
+        let s_m = state.s_m.clone();
+        let s_l = state.s_l.clone();
```
We should take these by reference.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/oracles.rs:
```diff
@@ -55,12 +55,22 @@ pub(in crate::snark::marlin) struct SingleEntry<'a, F: PrimeField> {
+    /// The lookup constraint selector polynomial.
+    pub(super) s_l_poly: LabeledPolynomial<F>,
+    /// The query vector polynomial.
+    pub(super) f_poly: LabeledPolynomial<F>,
```
`s_m_poly` and `s_l_poly` are preprocessed polynomials and should be part of the `Index`, not part of the prover polynomial oracles.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -74,36 +74,53 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         {
             job_pool.add_job(move || Self::calculate_w(witness_label("w", i), private_variables, x_poly, state_ref));
-            job_pool.add_job(move || Self::calculate_z_m(witness_label("z_a", i), z_a, false, state_ref, None));
+            let z_a_clone = z_a.clone();
```
Why was this clone necessary?

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -74,36 +74,53 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            job_pool.add_job(move || Self::calculate_z_m(witness_label("z_a", i), z_a_clone, false, state_ref, None));
             let r_b = F::rand(rng);
-            job_pool.add_job(move || Self::calculate_z_m(witness_label("z_b", i), z_b, true, state_ref, Some(r_b)));
+            let z_b_clone = z_b.clone();
```
Ditto

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -74,36 +74,53 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
             if MM::ZK {
                 r_b_s.push(r_b);
             }
+            let z_c_clone = z_c.clone();
```
Ditto

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -74,36 +74,53 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            let z_c_clone = z_c.clone();
+            job_pool
+                .add_job(move || Self::calculate_z_m(witness_label("z_c", i), z_c_clone, true, state_ref, Some(r_b)));
+            let s_l_clone = s_l.clone();
```
Same

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -238,12 +255,61 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        // let poly_time = start_timer!(|| format!("Computing {label}"));
+
+        let evaluations = z_a
+            .iter()
```
This can be made into a `cfg_iter`

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -238,12 +255,61 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        PoolResult::Query(LabeledPolynomial::new(label, poly, None, None))
+    }
+
+    fn calculate_selector<'a>(
```
This should be done during indexing

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -238,12 +255,61 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
     Witness(LabeledPolynomial<F>),
     MatrixPoly(LabeledPolynomial<F>, LabeledPolynomialWithBasis<'a, F>),
+    Query(LabeledPolynomial<F>),
+    Selector(LabeledPolynomial<F>),
```
Once we move the selector polynomial computation to indexing, we can get rid of this variant.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs:
```diff
@@ -260,4 +326,18 @@ impl<'a, F: PrimeField> PoolResult<'a, F> {
+            Self::Selector(poly) => Some(poly),
+            _ => None,
+        }
+    }
```
Ditto

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs:
```diff
@@ -193,7 +194,16 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+                    // turning this back into a simpler lincheck sumcheck to allow easy
+                    // modification of the rowcheck.
                     let mut summed_z_m = {
+                        cfg_iter_mut!(z_b.coeffs).for_each(|b| *b *= eta_b);
```
I suspect we can change this into a single expression like

`let summed_z_m = cfg_iter!(z_a.coeffs).zip(z_b.coeffs).zip(z_c.coeffs).map(|((a, b), c)| a + eta_b * b + eta_c * c).collect();`

jules replied:
So it turns out that `z_a` is actually 1 below the degree of `z_b` and`z_c`. Would simply pushing a `F::zero()` to the end of `z_a` work out in this case or could the degree mismatch fluctuate?

jules replied:
I've added a bit of code that will just fill up `z_a` with zero coefficients until it matches `z_b`, though maybe this might not work in the non-ZK case (since the degree of `z_b` is otherwise just the size of the constraint domain, but i'm assuming this could vary?)

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/state.rs:
```diff
@@ -68,6 +68,22 @@ pub struct State<'a, F: PrimeField, MM: MarlinMode> {
+    /// The length of this list must be equal to the batch size.
+    pub(super) z_c: Option<Vec<Vec<F>>>,
+
+    /// The list of multiplication selector vectors for each instance in the batch.
+    /// The length of this list must be equal to the batch size.
+    pub(super) s_m: Vec<Vec<F>>,
+
+    /// The list of lookup selector vectors for each instance in the batch.
+    /// The length of this list must be equal to the batch size.
+    pub(super) s_l: Vec<Vec<F>>,
```
Since we support only the same circuit for all entries in a batch, this is shared across the entire batch, and should live only in the index.

jules committed:
Simplify lincheck sumcheck expression

jules committed:
Use cfg_iter! when calculating f

Pratyush commented on algorithms/src/snark/plookup/table.rs:
```diff
@@ -0,0 +1,40 @@
+
+use snarkvm_fields::Field;
+
+pub struct LookupTable<F: Field>(Vec<Vec<F>>);
```
For any circuit of interest, I believe that (at least) the number of columns is statically known. So I believe we can change this to `LookupTable<F: Field, const C: usize>(Vec<[F; C]>);`, which should reduce the number of redirections when accessing this vector.

jules replied:
It makes sense - however, this might kind of get in the way when we want to have multiple lookup tables with different column amounts. If that's no issue, then I'll opt for this but otherwise it may be better to stay unconstrained in this regard.

jules replied:
Although from an API point of view it is a bit nicer to constrain here since we won't have to deal with edge cases that come with needing to take in a dynamic-sized vector of lookup values - I'm also assuming we may get some performance gain out of being able to leverage arrays.

In order to save on space, I assume I can just manually implement the serialization-deserialization process to ignore zero elements in tables where the arity is smaller than the constrained size. It will sadly still be inefficient in-memory, but maybe we can find another way around this later.

jules replied:
After quite a bit more experimentation, it seems that adding this const generic has an enormous cascading effect on the required generics for every single type related to calculating SNARKs, including the `SNARK` trait itself (as it uses `ConstraintSynthesizer`s which directly extend `ConstraintSystem`, which contains the `LookupTable<F, C>`).

If this is really desired, I can proceed on finalizing this work, but it has a bad design smell IMO. It may be more convenient to instead opt for an `Indexmap` which will just allow map-like lookups, and constrain key lengths internally.

jules committed:
Reduce cloning in first round

Pratyush commented on algorithms/src/snark/plookup/table.rs:
```diff
@@ -0,0 +1,40 @@
+
+
+
+    pub fn compress(&self, zeta: F) -> Vec<F> {
```
Where is this method called? It needs to be called in the prover algorithm, right? Also, for efficiency, it's probably better to take a vector of length `num_columns - 1`.

jules replied:
Yeah, it should be - this was left over from my first approach and I mostly neglected it as I started delving into the PIOP stuff. I'll use it and update the API so the caller provides a `Vec::with_capacity(num_columns - 1)`

jules replied:
On second thought - this method is actually quite useless. It serves a good purpose in PLONK-style proof systems since it generates the query vector but our query vector is actually contingent on the generation of the matrix-vector products when initializing the prover. It may make more sense to fully contain the lookup table within the R1CS crate and remove the plookup mod altogether.

jules committed:
Fix summed_z_m calculation

jules committed:
Selector polynomials are now part of indexer polynomials

jules committed:
Revert back to old way of summing z_m

We cant guarantee that the degrees between
the three polynomials are the same

jules committed:
Remove leftover println

jules committed:
Remove 'plookup' module

Due to the way Marlin works, theres little benefit to
separately housing plonkup logic and exporting it, since it most likely
will encumber the structured logic thats currently in place.

jules committed:
Update R1CS abstraction to support dynamic-size lookups and multiple tables

jules committed:
Update Marlin to conform to new trait signatures, address review comments

jules committed:
Fix clippy

jules committed:
Remove stray import

jules committed:
Return error on submitting a lookup key of the wrong length

jules committed:
Add de/serialization logic for LookupTable

jules committed:
Store circuit lookup tables in Circuit struct

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
 
+        let z_a_s_at_beta = z_a_s.iter().map(|z_a| evals.get_lc_eval(z_a, beta)).collect::<Result<Vec<_>, _>>()?;
         let z_b_s_at_beta = z_b_s.iter().map(|z_b| evals.get_lc_eval(z_b, beta)).collect::<Result<Vec<_>, _>>()?;
+        let z_c_s_at_beta = z_c_s.iter().map(|z_c| evals.get_lc_eval(z_c, beta)).collect::<Result<Vec<_>, _>>()?;
```
I think we do not need to have `z_c_s_at_beta`?

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         let v_H_at_beta = constraint_domain.evaluate_vanishing_polynomial(beta);
         let v_X_at_beta = input_domain.evaluate_vanishing_polynomial(beta);
 
+        let z_a_s_at_beta = z_a_s.iter().map(|z_a| evals.get_lc_eval(z_a, beta)).collect::<Result<Vec<_>, _>>()?;
```
We don't need this either.

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+        let z_c_s_at_beta = z_c_s.iter().map(|z_c| evals.get_lc_eval(z_c, beta)).collect::<Result<Vec<_>, _>>()?;
+        let f_s_at_beta = f_s.iter().map(|f| evals.get_lc_eval(f, beta)).collect::<Result<Vec<_>, _>>()?;
+
+        let batch_z_a_at_beta: F =
+            z_a_s_at_beta.iter().zip_eq(batch_combiners).map(|(z_a_at_beta, combiner)| *z_a_at_beta * combiner).sum();
```
```suggestion
```

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            z_a_s_at_beta.iter().zip_eq(batch_combiners).map(|(z_a_at_beta, combiner)| *z_a_at_beta * combiner).sum();
         let batch_z_b_at_beta: F =
             z_b_s_at_beta.iter().zip_eq(batch_combiners).map(|(z_b_at_beta, combiner)| *z_b_at_beta * combiner).sum();
+        let batch_z_c_at_beta: F =
+            z_c_s_at_beta.iter().zip_eq(batch_combiners).map(|(z_c_at_beta, combiner)| *z_c_at_beta * combiner).sum();
```
```suggestion
```

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         let mut linear_combinations = BTreeMap::new();
 
         // Lincheck sumcheck:
+        let z_a_s = (0..state.batch_size)
+            .map(|i| {
+                let z_a_i = witness_label("z_a", i);
```
I suggest instead of constructing the linear combination for the rowcheck as `(z_a_beta * z_b_beta - z_c_beta) * s_m`, we should construct it as `(z_a * z_b_beta - z_c) * s_m_beta`. This saves  proof size: for a batch size of `n`,we only need to provide 1 evaluation of `s_m_beta` and `n` evaluations of `z_b_beta`, as opposed to `n` evaluations each of `z_a_beta`, `z_b_beta`, and `z_c_beta`.

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -217,29 +245,51 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
                     .add(-t_at_beta * v_X_at_beta * combiner, witness_label("w", i));
             }
             lincheck_sumcheck
+                .add(r_alpha_at_beta * batch_z_a_at_beta, LCTerm::One)
                 .add(r_alpha_at_beta * eta_b * batch_z_b_at_beta, LCTerm::One)
+                .add(r_alpha_at_beta * eta_c * batch_z_c_at_beta, LCTerm::One)
```
These should be replaced with `sum(r_alpha_at_beta * eta_m * z_m)`

Pratyush commented on algorithms/src/snark/marlin/ahp/ahp.rs:
```diff
@@ -217,29 +245,51 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+            let mut lincheck_sumcheck = LinearCombination::empty("placeholder");
+            // Add rowcheck
+            let mut rowcheck = LinearCombination::empty("lincheck_sumcheck");
+            for (i, combiner) in batch_combiners.iter().enumerate() {
+            rowcheck
+                .add((z_a_s_at_beta[i] * z_b_s_at_beta[i] - z_c_s_at_beta[i]) * combiner, "s_m")
+                .add((z_a_s_at_beta[i] + zeta * z_b_s_at_beta[i] + zeta_squared
+                    * z_c_s_at_beta[i] - f_s_at_beta[i]) * combiner, "s_l");
```
I suggest instead of constructing the linear combination for the rowcheck as `(z_a_beta * z_b_beta - z_c_beta) * s_m`, we should construct it as `(z_a * z_b_beta - z_c) * s_m_beta`. This saves  proof size: for a batch size of `n`,we only need to provide 1 evaluation of `s_m_beta` and `n` evaluations of `z_b_beta`, as opposed to `n` evaluations each of `z_a_beta`, `z_b_beta`, and `z_c_beta`. This results in a savings of ~2n field elements.

Similarly for the lookup part, we should do it as `(z_a + zeta * z_b + zeta * z_c) * s_l_beta`.

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs:
```diff
@@ -123,7 +141,18 @@ impl<F: Field> CS<F> for ConstraintSystem<F> {
+    }
+
+    fn lookup(&mut self, _: &[LinearCombination<F>], table_index: usize) -> Result<Variable, SynthesisError> {
+        self.a.push(Self::make_row(&LinearCombination::zero()));
```
Shouldn't the various `self.m.push(...)` be identical to `enforce` above?

jules replied:
yep, good catch! seems i was just pushing zeros this whole time

Pratyush commented on algorithms/src/snark/marlin/ahp/indexer/constraint_system.rs:
```diff
@@ -123,7 +141,18 @@ impl<F: Field> CS<F> for ConstraintSystem<F> {
+        self.num_constraints += 1;
+    }
+
+    fn lookup(&mut self, _: &[LinearCombination<F>], table_index: usize) -> Result<Variable, SynthesisError> {
```
I think the API for this should be similar to the one for `enforce`: we should take in three linear combinations `a, b, c`, and check that the table contains the triple `(az, bz, cz)`. So the API would look like `enforce_lookup(a, b, c)`.

We could additionally keep the existing API of `c = lookup(a, b)` as sugar atop the core API.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/constraint_system.rs:
```diff
@@ -56,11 +70,34 @@ impl<F: Field> ConstraintSystem<F> {
         assert_eq!(self.num_public_variables + self.num_private_variables, self.num_constraints, "padding failed!");
     }
+
+    fn lookup(&mut self, key: &[LinearCombination<F>], table_index: usize) -> Result<F, SynthesisError> {
```
To confirm, this is a "native" lookup, right?

jules replied:
yeah!

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/mod.rs:
```diff
@@ -30,21 +30,22 @@ use snarkvm_utilities::println;
 #[cfg(feature = "parallel")]
 use rayon::prelude::*;
 
+mod fifth;
 mod first;
 mod fourth;
 mod second;
 mod third;
```
Nit: let's sort this by round (if rustfmt re-orders things, let's use a `#[rustfmt::skip]`)

jules replied:
to do this we'd either have to bury them in a separate mod (so that we can scope the directive), or tag each line with `#[rustfmt::skip]` - if there is no problem adding an extra mod layer then i'll go for that!

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -14,222 +14,89 @@
-        let oracles = prover::ThirdOracles { g_a, g_b, g_c };
-        state.lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
-        state.sums = Some([sum_a, sum_b, sum_c]);
+            .expect("prover::State should include verifier_first_msg when prover_fourth_round is called");
```
```suggestion
            .expect("prover::State should include verifier_first_msg when prover_third_round is called");
```

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs:
```diff
@@ -14,222 +14,89 @@
+            })
+            .sum::<DensePolynomial<F>>();
+
+        let mut lin = state.lin.as_ref().unwrap().clone();
+        cfg_iter_mut!(lin.coeffs).for_each(|c| *c *= delta);
+        let mut x_g_1 = state.x_g_1.as_ref().unwrap().clone();
+        cfg_iter_mut!(x_g_1.coeffs).for_each(|c| *c *= delta);
+
+        let q_1 = &(&row + &lin) - &x_g_1;
+        let (h_1, rem) = q_1.divide_by_vanishing_poly(constraint_domain).unwrap();
```
We can just save `_h_1` from the previous round and use it here.

Pratyush commented on algorithms/src/snark/marlin/ahp/prover/state.rs:
```diff
@@ -83,6 +91,12 @@ pub struct State<'a, F: PrimeField, MM: MarlinMode> {
     /// The challenges sent by the verifier in the first round
     pub(super) verifier_first_message: Option<verifier::FirstMessage<F>>,
 
+    /// The lincheck sumcheck polynomial produced in the second round.
+    pub(super) lin: Option<DensePolynomial<F>>,
+
+    /// The Xg_1(X) polynomial produced in the second round.
+    pub(super) x_g_1: Option<DensePolynomial<F>>,
```
We *should* be able to replace this with `_h_1`

jules committed:
Modify lookup table to always be 3-arity

Since we only have 3 matrices, we can not enforce larger lookups

jules committed:
Interpolate and index t poly

jules committed:
Interpolate s polys

jules committed:
Incorporate s polys into first round oracles

jules committed:
t poly is now same length as num_constraints

jules committed:
Label s polys correctly

jules committed:
Enter correct evaluation in case selector poly is zero

jules committed:
Fix typo

jules committed:
Ensure correct chunking during proof generation

jules committed:
Use h_1 generated in second round

jules committed:
Add z_2 poly computation

jules committed:
Compute quotient checks for lookup poly

jules committed:
Ensure lookup poly is calculated within constraint bounds

The lookup poly was instead calculated in the domain bound
and, due to unexpected zero values, produced an inconsistent result
which caused a remainder to exist after dividing by h_1

julesdesmit commented:
NOTE: a lot of the code currently being pushed is heavily unoptimized - this will change in the near future once the protocol itself is functioning and I can redirect my focus towards cutting corners. Leaving it in this state currently makes it a little easier to reason with since it's more verbose.

jules committed:
Avoid segfaults when populating t_evals

jules committed:
Reduce evaluations in proof, clean up linear combination construction

jules committed:
Simplify t poly calculation

jules committed:
Optimize lookup quotient poly calculation

jules committed:
Update R1CS API for lookups

jules committed:
Fixes for plookup protocol

jules committed:
Remove unnecessary padding of f_evals

jules committed:
Add plookup terms into linear combination construction

Tests are still failing

jules committed:
Formatting, reduce cloning

jules committed:
Use correct evals for interpolation of delta_t_omega

jules committed:
Lookup terms pass

jules committed:
Remove excessive cloning, remove storage of plookup poly evals

jules committed:
Fix failing test cases where table is empty

jules committed:
Keep l_1 on the index instead of interpolating it on the fly

jules committed:
Add comment

jules committed:
Canonical ordering of selectors and table polys

jules committed:
Remove comment

Since the domain of the l_1 poly still depends on the circuit,
it makes sense to keep this on the index.

jules committed:
Restore renaming

jules committed:
Derive randomizers correctly, cleanup

jules committed:
Minor fixes

jules committed:
Fix sponge ordering, fix typos

jules committed:
Revert more leftover renaming

jules committed:
Revert extraction of interpolation function from first round

jules committed:
Precompute multiplications by delta for s_1_omega

This cuts a small corner on performance

jules committed:
Use itertools in second round

jules commented:
This work should now be mostly wrapped up, cleaned and optimized for as far as I can manage naively. Next up to finish this will be to make it fit back into the CI properly (some integration stuff fails because proving keys have changed and they will no longer deserialize properly), but since there is ongoing work on the parameter infrastructure, I'll leave this PR to be revisited when appropriate.

jules committed:
Merge branch 'testnet3' into marlin_lookups

jules committed:
Fix clippy
